File: app/web/src/api/mute.ts
import apiClient from "./apiClient";

export interface MuteRecord {
	id: number;
	uuid: string;
	user_id: number;
	muter_id: number;
	duration: number;
	permanent: boolean;
	expires_at: string | null;
	reason: string;
	created_at: string;
	updated_at: string;
}

export interface CreateMuteParams {
	user_id: number;
	duration: number;
	permanent: boolean;
	reason: string;
}

export async function createMute(
	params: CreateMuteParams,
): Promise<MuteRecord> {
	const data = await apiClient.post<MuteRecord>({
		url: "/api/v1/mute/create",
		data: params,
	});

	if (!data) throw new Error("mute record is missing");
	return data;
}

export async function cancelMute(userId: number): Promise<void> {
	await apiClient.post({
		url: "/api/v1/mute/cancel",
		data: { user_id: userId },
	});
}

export async function listMutes(): Promise<MuteRecord[]> {
	const data = await apiClient.post<MuteRecord[]>({
		url: "/api/v1/mute/list",
	});

	return data || [];
}

File: app/server/internal/model/mute.go
package model

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Mute 持久化禁言记录，全局生效
type Mute struct {
	ID        uint       `gorm:"primaryKey" json:"id"`
	UUID      string     `gorm:"type:uuid;uniqueIndex" json:"uuid"`
	UserID    uint       `gorm:"uniqueIndex" json:"user_id"`
	MuterID   uint       `json:"muter_id"`
	Duration  int64      `json:"duration"` // 禁言秒数，0 = 永久
	Permanent bool       `gorm:"default:false" json:"permanent"`
	ExpiresAt *time.Time `json:"expires_at"` // nil = 永久禁言
	Reason    string     `gorm:"size:255" json:"reason"`
	CreatedAt time.Time  `json:"created_at"`
	UpdatedAt time.Time  `json:"updated_at"`
}

func (Mute) TableName() string {
	return "mutes"
}

func (m *Mute) BeforeCreate(_ *gorm.DB) error {
	if m.UUID == "" {
		m.UUID = uuid.New().String()
	}
	return nil
}
